# Lab | Intro to Probability

**Objective**

Welcome to this Intro to Probability lab, where we explore decision-making scenarios through the lens of probability and strategic analysis. In the business world, making informed decisions is crucial, especially when faced with uncertainties. This lab focuses on scenarios where probabilistic outcomes play a significant role in shaping strategies and outcomes. Students will engage in exercises that require assessing and choosing optimal paths based on data-driven insights. The goal is to enhance your skills by applying probability concepts to solve real-world problems.

**Challenge 1**

#### Ironhack Airlines 

Often Airlines sell more tickets than they have seats available, this is called overbooking. Consider the following:
- A plane has 450 seats. 
- Based on historical data we conclude that each individual passenger has a 3% chance of missing it's flight. 

If the Ironhack Airlines routinely sells 460 tickets, what is the chance that they have a seats for all passenger?

In [1]:
from scipy.stats import binom

In [2]:
# Binomial

n = 460        # tickets sold
p_show = 0.97  # show-up probability
seats = 450    # plane capacity

prob_enough_seats = binom.cdf(seats, n, p_show)
print(f"Probability there are seats for everyone: {prob_enough_seats:.4%}")

Probability there are seats for everyone: 88.4477%


The chance that they have a seat for all passenger is 88.45%

**Challenge 2**

#### Ironhack Call Center 

Suppose a customer service representative at a call center is handling customer complaints. Consider the following:
- The probability of successfully resolving a customer complaint on the first attempt is 0.3. 


What is the probability that the representative needs to make at least three attempts before successfully resolving a customer complaint?

In [3]:
from scipy.stats import geom

In [ ]:
# Geometric
p = 0.3

prob_at_least_3 = geom.sf(2, p)   
print(f"Probability needs ≥3 attempts: {prob_at_least_3:.2%}") 

Probability needs ≥3 attempts: 49.00%


In [5]:
# or
prob_at_least_3_v2 = 1 - geom.cdf(2, p)
print(f"Probability needs ≥3 attempts (v2): {prob_at_least_3_v2:.2%}")

Probability needs ≥3 attempts (v2): 49.00%


In [6]:
# or
prob_at_least_3_v3 = (1 - p) ** 2
print(f"Probability needs ≥3 attempts (v3): {prob_at_least_3_v3:.2%}")

Probability needs ≥3 attempts (v3): 49.00%


**Challenge 3**

#### Ironhack Website

Consider a scenario related to Ironhack website traffic. Where:
- our website takes on average 500 visits per hour.
- the website's server is designed to handle up to 550 vists per hour.


What is the probability of the website server being overwhelmed?

In [7]:
from scipy.stats import poisson

In [8]:
# Poisson

lam = 500
capacity = 550

p_hour = poisson.sf(capacity, lam) 
print(f"Probability overwhelmed in one hour: {p_hour:.2%}") 

Probability overwhelmed in one hour: 1.29%


What is the probability of being overwhelmed at some point during a day? (consider 24hours)

In [9]:
p_day = 1 - (1 - p_hour)**24
print(f"Probability overwhelmed at least once in 24h: {p_day:.2%}")

Probability overwhelmed at least once in 24h: 26.77%


**Challenge 4**

#### Ironhack Helpdesk

Consider a scenario related to the time between arrivals of customers at a service desk.

On average, a customers arrives every 10minutes.

What is the probability that the next customer will arrive within the next 5 minutes?

In [10]:
from scipy.stats import expon

In [13]:
# Exponential

rate = 1/10  # per minute
scale = 1/rate  

p_within_5 = expon.cdf(5, scale=scale) 
print(f"Next customer arrives within 5 min: {p_within_5:.2%}") 

Next customer arrives within 5 min: 39.35%


If there is no customer for 15minutes, employees can that a 5minutes break.

What is the probability an employee taking a break?

In [15]:
p_start_break = expon.sf(15, scale=scale) 
p_finish_break = expon.sf(20, scale=scale)  
    
print(f"Can start a break in: {p_start_break:.2%}")     

Can start a break in: 22.31%


**Challenge 5**

The weights of a certain species of birds follow a normal distribution with a mean weight of 150 grams and a standard deviation of 10 grams. 

- If we randomly select a bird, what is the probability that its weight is between 140 and 160 grams?

In [16]:
from scipy.stats import norm

In [18]:
# Normal

mu, sigma = 150, 10
prob_140_160 = norm.cdf(160, mu, sigma) - norm.cdf(140, mu, sigma)
print(f"Probability of weight (140g ≤ weight ≤ 160g): {prob_140_160:.2%}")

Probability of weight (140g ≤ weight ≤ 160g): 68.27%


**Challenge 6**

If the lifetime (in hours) of a certain electronic component follows an exponential distribution with a mean lifetime of 50 hours, what is the probability that the component fails within the first 30 hours?

In [19]:
from scipy.stats import expon

In [20]:
# Exponential

mean_hours = 50
t = 30

p_fail_30 = expon.cdf(t, scale=mean_hours)
print(f"Probability it fails within first {t} hours: {p_fail_30:.2%}")

Probability it fails within first 30 hours: 45.12%
